src/clp_ffi_js/ir/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,233 @@
+#include "ClpIrV1Decoder.hpp"
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <span>
+#include <string>
+#include <system_error>
+#include <utility>
+
+#include <clp/ErrorCode.hpp>
+#include <clp/ffi/encoding_methods.hpp>
+#include <clp/ffi/ir_stream/decoding_methods.hpp>
+#include <clp/ir/LogEvent.hpp>
+#include <clp/ir/LogEventDeserializer.hpp>
+#include <clp/ir/types.hpp>
+#include <clp/streaming_compression/zstd/Decompressor.hpp>
+#include <clp/TraceableException.hpp>
+#include <emscripten/em_asm.h>
+#include <emscripten/val.h>
+#include <spdlog/spdlog.h>
+
+#include <clp_ffi_js/ClpJsException.hpp>
+#include <clp_ffi_js/constants.hpp>
+
+using namespace std::literals::string_literals;
+
+namespace clp_ffi_js::ir {
+auto ClpIrV1Decoder::create(emscripten::val const& data_array) -> ClpIrV1Decoder {
+    auto const length{data_array["length"].as<size_t>()};
+    SPDLOG_INFO("ClpIrV1Decoder::ClpIrV1Decoder() got buffer of length={}", length);
+
+    auto data_buffer{std::make_unique<char const[]>(length)};
+    emscripten::val::module_property("HEAPU8")
+            .call<void>("set", data_array, reinterpret_cast<uintptr_t>(data_buffer.get()));
+
+    auto const zstd_decompressor{std::make_shared<clp::streaming_compression::zstd::Decompressor>()
+    };
+    zstd_decompressor->open(data_buffer.get(), length);
+
+    bool is_four_bytes_encoding{true};
+    if (auto const err{
+                clp::ffi::ir_stream::get_encoding_type(*zstd_decompressor, is_four_bytes_encoding)
+        };
+        clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success != err)
+    {
+        SPDLOG_CRITICAL("Failed to decode encoding type, err={}", err);
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decode encoding type."
+        );
+    }
+
+    if (false == is_four_bytes_encoding) {
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_Unsupported,
+                __FILENAME__,
+                __LINE__,
+                "Is not four byte encoding."
+        );
+    }
+    auto result{clp::ir::LogEventDeserializer<clp::ir::four_byte_encoded_variable_t>::create(
+            *zstd_decompressor
+    )};
+    if (result.has_error()) {
+        auto const error_code{result.error()};
+        SPDLOG_CRITICAL(
+                "Failed to decompress: {}:{}",
+                error_code.category().name(),
+                error_code.message()
+        );
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
```

Comment:
```suggestion
                clp::ErrorCode::ErrorCode_Failure,
```

I could see how a truncated IR stream (one of the possible errors from `create`) might be interpreted as `MetadataCorrupted`, but it may be misleading for someone who doesn't read `create`'s docstring.